config: handover matcher settings to claircore

arajkumar opened the pull request:
A root level config key `matchers` is added to customize the `Matcher` construction.

Signed-off-by: Arunprasad Rajkumar <arajkuma@redhat.com>

ldelossa reviewed (changes requested):
@arajkumar 

Do we need to make a distinction between "RemoteMatcher" configs and general "Matcher" configs? 

I also do not think we need "RemoteMatcher" as top level global key, when a "Matcher" stanza already exists. Can you find a way to do this config inside the "Matcher" stanza and possible make the config agnostic to whether its a "Remote" or "Local" matcher?

arajkumar commented:
> 
> Do we need to make a distinction between "RemoteMatcher" configs and general "Matcher" configs?

IMO, normal Matcher hardly needs configurability. Do you have any use case for that?
> 
> I also do not think we need "RemoteMatcher" as top level global key, when a "Matcher" stanza already exists. Can you find a way to do this config inside the "Matcher" stanza and possible make the config agnostic to whether its a "Remote" or "Local" matcher?

This is doable.

ldelossa commented:
We often get requests to block specific Vulns or packages from matching,
this is a good opportunity to allow this configuration into matchers.

On Wed, Feb 24, 2021 at 2:18 AM Arunprasad Rajkumar <
notifications@github.com> wrote:

> Do we need to make a distinction between "RemoteMatcher" configs and
> general "Matcher" configs?
>
> IMO, normal Matcher hardly needs configurability. Do you have any use case
> for that?
>
> I also do not think we need "RemoteMatcher" as top level global key, when
> a "Matcher" stanza already exists. Can you find a way to do this config
> inside the "Matcher" stanza and possible make the config agnostic to
> whether its a "Remote" or "Local" matcher?
>
> This is doable.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/quay/clair/pull/1187#issuecomment-784851797>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLBVFUXQOCUCR77JRSSSETTASR43ANCNFSM4XW3G2BA>
> .
>

arajkumar commented:
@ldelossa 
>We often get requests to block specific Vulns or packages from matching,

Isn't this come under policy purview? This shall be done at consumer level,.e.g registry org admin shall have a policy to ignore vulns/pkgs which is not affecting their use case.

ldelossa commented:
Dont think so.

When we implement proper registry all registry components will get config.
No need to special case things.

On Wed, Feb 24, 2021 at 12:05 PM Arunprasad Rajkumar <
notifications@github.com> wrote:

> @ldelossa <https://github.com/ldelossa>
>
> We often get requests to block specific Vulns or packages from matching,
>
> Isn't this come under policy purview? This shall be done at consumer
> level,.e.g registry org admin shall have a policy to ignore vulns/pkgs
> which is not affecting their use case.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/quay/clair/pull/1187#issuecomment-785226658>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLBVFRHMUJEQS7KUGEF5Q3TAUWV5ANCNFSM4XW3G2BA>
> .
>

arajkumar commented:
@ldelossa , updated  `config.yaml.sample` which moves matcher configurability under `matcher` stanza. Let's first agree upon a config format, I will make code changes later.

arajkumar commented:
>> Isn't this come under policy purview? This shall be done at consumer level,.e.g registry org admin shall have a policy to ignore vulns/pkgs which is not affecting their use case.

> When we implement proper registry all registry components will get config.

I actually meant container registry.

for example, consider a container image which has a known vuln, but the admin of the container registry org aware that the vuln is not affecting their use case. Then the admin can take the decision to mute a specific vuln/pkg. I'm not sure whether this option is already present in any of the popular registry hubs(quay.io, docker, etc).

ldelossa commented on config.yaml.sample:
```diff
@@ -21,6 +21,9 @@ indexer:
   layer_scan_concurrency: 5
   migrations: true
 matcher:
+  matchers:
```
@arajkumar 

Lets use the way we currently plug "Updater" config into clair as an example: https://quay.github.io/clair/reference/config.html#updaters-object

Can we emulate this same structure with a "Matcher" and "Config" subkey? 

It would be nice to keep things looking the same. @hdonnay please chime in asap if you do not agree.

arajkumar replied:
@ldelossa If I understand correctly, it will look like below,

```
matcher:
    connstring: ""
    max_conn_pool: 0
    indexer_addr: ""
    migrations: false
    period: ""
    disable_updaters: false
    update_retention: 2
    sets:
        - rhel
        - ubuntu
        - python
        - ..
        - crda
    config:
       crda:
           host: https://blabla
           concurrency_limit: 10
````

Sounds good to me.

ldelossa replied:
not exactly. Place Matchers block at root of config like updaters is currently. This was a decision to signify multiple services may use that block, not just the matcher (for instance, the omnimatcher in the indexer might)

```
matchers:
  names: (since we don't have a "set" concept for matchers)
    - rhel
    - ubuntu
    - python
    - ...
    - crda
  config:
    crda:
        ...

arajkumar commented:
Changes are made in sync with https://github.com/quay/claircore/pull/337

arajkumar committed:
config: add matchers settings

A root level config key matchers is added to customize the Matcher construction.

Signed-off-by: Arunprasad Rajkumar <arajkuma@redhat.com>

arajkumar commented on go.mod:
```diff
@@ -16,24 +17,24 @@ require (
 	github.com/prometheus/client_golang v1.9.0
 	github.com/prometheus/procfs v0.3.0 // indirect
-	github.com/quay/claircore v0.3.1
+	github.com/quay/claircore v0.3.2-0.20210316145942-25dd763a32d2
```
Just a temporary measure until we get a next semver release.

ldelossa reviewed (changes requested):
Hey Arun, 

This needs config reference doc updates: https://github.com/quay/clair/blob/main/Documentation/reference/config.md

Other then this, we need to keep in mind that CRDA should *not* wind up in the omnimatcher. What you have is fine, but make sure when you wire this all together, this list of default matcher names do not get auto-injected into the omni-matcher.

arajkumar commented:
@ldelossa 
>This needs config reference doc updates: https://github.com/quay/clair/blob/main/Documentation/reference/config.md

Done.

>Other then this, we need to keep in mind that CRDA should not wind up in the omnimatcher.

I'll make sure this never happens.

ldelossa reviewed (approved):
@hdonnay cc - a bit more config updates if you plan on rebaing jsonPath config commits you have. Take note on these new ones.
